File: scripts/analyze_patterns.py
#!/usr/bin/env python3
"""
Analyze and compare results from different load pattern tests.
Usage: python3 analyze_patterns.py <results_file.jtl> [results_file2.jtl ...]
"""

import csv
import json
import sys
from pathlib import Path
from collections import defaultdict
import statistics

def parse_jtl(filepath):
    """Parse JMeter JTL CSV results file."""
    results = {
        'samples': [],
        'stats': defaultdict(list)
    }
    
    try:
        with open(filepath, 'r') as f:
            reader = csv.DictReader(f)
            for row in reader:
                results['samples'].append(row)
    except FileNotFoundError:
        print(f"Error: File not found: {filepath}")
        return None
    
    return results

def analyze_results(results, name):
    """Analyze JTL results and calculate metrics."""
    if not results or not results['samples']:
        print(f"No samples found in {name}")
        return None
    
    samples = results['samples']
    response_times = []
    latencies = []
    success_count = 0
    error_count = 0
    
    for sample in samples:
        try:
            # JMeter columns: timeStamp, elapsed, label, responseCode, responseMessage, threadName, dataType, success, failureMessage, bytes, sentBytes, grpThreads, allThreads, Latency, IdleTime, Connect
            elapsed = int(sample.get('elapsed', 0))
            success = sample.get('success', 'false').lower() == 'true'
            latency = int(sample.get('Latency', elapsed))
            
            response_times.append(elapsed)
            latencies.append(latency)
            
            if success:
                success_count += 1
            else:
                error_count += 1
        except (ValueError, KeyError):
            continue
    
    if not response_times:
        return None
    
    total_samples = success_count + error_count
    error_rate = (error_count / total_samples * 100) if total_samples > 0 else 0
    
    metrics = {
        'name': name,
        'total_samples': total_samples,
        'success': success_count,
        'errors': error_count,
        'error_rate_pct': error_rate,
        'response_time': {
            'min': min(response_times),
            'max': max(response_times),
            'avg': statistics.mean(response_times),
            'median': statistics.median(response_times),
            'p95': sorted(response_times)[int(len(response_times) * 0.95)] if len(response_times) > 0 else 0,
            'p99': sorted(response_times)[int(len(response_times) * 0.99)] if len(response_times) > 0 else 0,
            'stdev': statistics.stdev(response_times) if len(response_times) > 1 else 0,
        },
        'latency': {
            'min': min(latencies),
            'max': max(latencies),
            'avg': statistics.mean(latencies),
            'median': statistics.median(latencies),
            'p95': sorted(latencies)[int(len(latencies) * 0.95)] if len(latencies) > 0 else 0,
            'p99': sorted(latencies)[int(len(latencies) * 0.99)] if len(latencies) > 0 else 0,
        },
        'throughput': total_samples / (len(samples) / 1000) if samples else 0,  # Rough estimate
    }
    
    return metrics

def print_metrics(metrics):
    """Pretty print metrics."""
    print(f"\n{'='*70}")
    print(f"📊 TEST: {metrics['name']}")
    print(f"{'='*70}")
    
    print(f"\n✅ Sample Count:")
    print(f"   Total:       {metrics['total_samples']:,} samples")
    print(f"   Success:     {metrics['success']:,}")
    print(f"   Errors:      {metrics['errors']:,}")
    print(f"   Error Rate:  {metrics['error_rate_pct']:.2f}%")
    
    print(f"\n⏱️  Response Time (ms):")
    rt = metrics['response_time']
    print(f"   Min:         {rt['min']:>10} ms")
    print(f"   Avg:         {rt['avg']:>10.2f} ms")
    print(f"   Median:      {rt['median']:>10} ms")
    print(f"   p95:         {rt['p95']:>10} ms")
    print(f"   p99:         {rt['p99']:>10} ms")
    print(f"   Max:         {rt['max']:>10} ms")
    print(f"   StdDev:      {rt['stdev']:>10.2f} ms")
    
    print(f"\n🔃 Latency (ms):")
    lat = metrics['latency']
    print(f"   Min:         {lat['min']:>10} ms")
    print(f"   Avg:         {lat['avg']:>10.2f} ms")
    print(f"   Median:      {lat['median']:>10} ms")
    print(f"   p95:         {lat['p95']:>10} ms")
    print(f"   p99:         {lat['p99']:>10} ms")
    print(f"   Max:         {lat['max']:>10} ms")
    
    print(f"\n📈 Throughput:")
    print(f"   ~{metrics['throughput']:.2f} requests/sec")

def compare_metrics(all_metrics):
    """Compare multiple test results."""
    if len(all_metrics) < 2:
        return
    
    print(f"\n\n{'='*70}")
    print("📋 COMPARISON SUMMARY")
    print(f"{'='*70}\n")
    
    print(f"{'Test Name':<25} {'Samples':<12} {'Error %':<10} {'Avg RT':<12} {'p95 RT':<12}")
    print(f"{'-'*70}")
    
    for m in all_metrics:
        print(f"{m['name']:<25} {m['total_samples']:<12,} "
              f"{m['error_rate_pct']:<10.2f} {m['response_time']['avg']:<12.1f} "
              f"{m['response_time']['p95']:<12}")
    
    print(f"\n💡 Key Insights:")
    
    # Find best/worst performers
    best_response = min(all_metrics, key=lambda x: x['response_time']['avg'])
    worst_response = max(all_metrics, key=lambda x: x['response_time']['avg'])
    worst_errors = max(all_metrics, key=lambda x: x['error_rate_pct'])
    
    print(f"   Best Response Time:  {best_response['name']} ({best_response['response_time']['avg']:.1f}ms avg)")
    print(f"   Worst Response Time: {worst_response['name']} ({worst_response['response_time']['avg']:.1f}ms avg)")
    print(f"   Highest Errors:      {worst_errors['name']} ({worst_errors['error_rate_pct']:.2f}%)")

def main():
    if len(sys.argv) < 2:
        print("Usage: python3 analyze_patterns.py <results.jtl> [results2.jtl ...]")
        print("\nExample:")
        print("  python3 scripts/analyze_patterns.py results-spike-*.jtl")
        print("  python3 scripts/analyze_patterns.py results-spike-*.jtl results-soak-*.jtl")
        sys.exit(1)
    
    all_metrics = []
    
    for filepath in sys.argv[1:]:
        print(f"📂 Parsing: {filepath}")
        results = parse_jtl(filepath)
        
        if results:
            metrics = analyze_results(results, Path(filepath).stem)
            if metrics:
                print_metrics(metrics)
                all_metrics.append(metrics)
    
    if all_metrics:
        compare_metrics(all_metrics)
    
    print("\n")

if __name__ == '__main__':
    main()

File: scripts/analyze_patterns.js
#!/usr/bin/env node

// analyze_patterns.js
// Simple JS script to analyze JMeter .jtl (CSV) results files.
// Usage: node scripts/analyze_patterns.js results-*.jtl

const fs = require('fs');
const path = require('path');

function parseCsv(filePath) {
  const raw = fs.readFileSync(filePath, 'utf8');
  const lines = raw.trim().split(/\r?\n/);
  if (lines.length < 2) return [];

  const headers = lines[0].split(',');
  const rows = lines.slice(1).map((line) => {
    const values = line.split(',');
    const obj = {};
    headers.forEach((header, i) => {
      obj[header] = values[i] === undefined ? '' : values[i];
    });
    return obj;
  });

  return rows;
}

function toNumber(value, fallback = 0) {
  const n = Number(value);
  return Number.isFinite(n) ? n : fallback;
}

function percentile(sorted, p) {
  if (!sorted.length) return 0;
  const idx = Math.min(sorted.length - 1, Math.floor(sorted.length * p));
  return sorted[idx];
}

function analyze(filePath) {
  const samples = parseCsv(filePath);
  if (!samples.length) {
    console.warn(`No data found in ${filePath}`);
    return null;
  }

  const responseTimes = [];
  const latencies = [];
  let successCount = 0;
  let errorCount = 0;

  for (const sample of samples) {
    const elapsed = toNumber(sample.elapsed, 0);
    const success = String(sample.success).toLowerCase() === 'true';
    const latency = toNumber(sample.Latency, elapsed);

    responseTimes.push(elapsed);
    latencies.push(latency);

    if (success) successCount += 1;
    else errorCount += 1;
  }

  responseTimes.sort((a, b) => a - b);
  latencies.sort((a, b) => a - b);

  const total = successCount + errorCount;
  const errorRate = total ? (errorCount / total) * 100 : 0;

  const stats = {
    totalSamples: total,
    success: successCount,
    errors: errorCount,
    errorRatePct: errorRate,
    responseTime: {
      min: responseTimes[0] ?? 0,
      max: responseTimes[responseTimes.length - 1] ?? 0,
      avg: responseTimes.reduce((a, b) => a + b, 0) / (responseTimes.length || 1),
      median: percentile(responseTimes, 0.5),
      p95: percentile(responseTimes, 0.95),
      p99: percentile(responseTimes, 0.99),
    },
    latency: {
      min: latencies[0] ?? 0,
      max: latencies[latencies.length - 1] ?? 0,
      avg: latencies.reduce((a, b) => a + b, 0) / (latencies.length || 1),
      median: percentile(latencies, 0.5),
      p95: percentile(latencies, 0.95),
      p99: percentile(latencies, 0.99),
    },
  };

  return stats;
}

function printStats(name, stats) {
  if (!stats) return;

  const fmt = (n) => `${Math.round(n * 100) / 100}`;

  console.log('\n' + '='.repeat(60));
  console.log(`RESULT: ${name}`);
  console.log('='.repeat(60));

  console.log('Total Samples: ', stats.totalSamples);
  console.log('Success:       ', stats.success);
  console.log('Errors:        ', stats.errors);
  console.log('Error Rate:    ', fmt(stats.errorRatePct) + '%');

  console.log('\nResponse Time (ms):');
  console.log('  min:  ', stats.responseTime.min);
  console.log('  avg:  ', fmt(stats.responseTime.avg));
  console.log('  p95:  ', stats.responseTime.p95);
  console.log('  p99:  ', stats.responseTime.p99);
  console.log('  max:  ', stats.responseTime.max);

  console.log('\nLatency (ms):');
  console.log('  min:  ', stats.latency.min);
  console.log('  avg:  ', fmt(stats.latency.avg));
  console.log('  p95:  ', stats.latency.p95);
  console.log('  p99:  ', stats.latency.p99);
  console.log('  max:  ', stats.latency.max);
}

function main() {
  const args = process.argv.slice(2);
  // If the user provides no arguments, automatically search for results-*.jtl files.
  const patterns = args.length ? args : ['results-*.jtl'];

  const reports = [];

  for (const filePattern of patterns) {
    const resolved = filePattern.includes('*')
      ? fs.readdirSync(process.cwd()).filter((f) => f.match(new RegExp(filePattern.replace('*', '.*'))))
      : [filePattern];

    for (const filePath of resolved) {
      if (!fs.existsSync(filePath)) continue;
      const stats = analyze(filePath);
      printStats(path.basename(filePath), stats);
      reports.push({ name: path.basename(filePath), stats });
    }
  }

  if (reports.length > 1) {
    console.log('\n' + '='.repeat(60));
    console.log('COMPARISON:');
    console.log('='.repeat(60));
    console.log('Name'.padEnd(30), 'Error %'.padEnd(10), 'Avg RT'.padEnd(10), 'p95 RT'.padEnd(10));

    reports.forEach((report) => {
      const s = report.stats;
      const name = report.name.padEnd(30);
      const err = fmt(s.errorRatePct).padEnd(10);
      const avg = fmt(s.responseTime.avg).padEnd(10);
      const p95 = String(s.responseTime.p95).padEnd(10);
      console.log(name, err, avg, p95);
    });
  }
}

main();
